src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -1421,18 +1421,115 @@ private static List<CompletionResult> CompletePropertyAssignment(MemberExpressio
         {
             if (SafeExprEvaluator.TrySafeEval(memberExpression, context.ExecutionContext, out var evalValue))
             {
-                if (evalValue is null)
+                if (evalValue is not null)
                 {
+                    Type type = evalValue.GetType();
+                    if (type.IsEnum)
+                    {
+                        return GetResultForEnum(type, context);
+                    }
+
                     return null;
                 }
+            }
 
-                Type type = evalValue.GetType();
-                if (type.IsEnum)
+            _ = TryGetInferredCompletionsForAssignment(memberExpression, context, out List<CompletionResult> result);
+            return result;
+        }
+
+        private static bool TryGetInferredCompletionsForAssignment(Ast expression, CompletionContext context, out List<CompletionResult> result)
+        {
+            result = null;
+            IList<PSTypeName> inferredTypes;
+            if (expression.Parent is ConvertExpressionAst convertExpression)
+            {
+                inferredTypes = new PSTypeName[] { new(convertExpression.Type.TypeName) };
+            }
+            else if (expression is MemberExpressionAst)
+            {
+                inferredTypes = AstTypeInference.InferTypeOf(expression);
+            }
+            else if (expression is VariableExpressionAst varExpression)
+            {
+                PSTypeName typeConstraint = CompletionCompleters.GetLastDeclaredTypeConstraint(varExpression, context.TypeInferenceContext);
+                if (typeConstraint is null)
                 {
-                    return GetResultForEnum(type, context);
+                    return false;
                 }
+
+                inferredTypes = new PSTypeName[] { typeConstraint };
             }
-            return null;
+            else
+            {
+                return false;
+            }
+             
+            if (inferredTypes.Count == 0)
+            {
+                return false;
+            }
+
+            var values = new SortedSet<string>();
```

Comment:
It is not mandatory request and it is not critical for the code but it looks sloppy if we don't follow the rule - return null if there are no values, otherwise return non-empty collection.
I'd move the initialization in line 1481 as `values ??= new SortedSet<string>();`
I think about this when I see a null check in the calling code.

Reply:
>  return null if there are no values, otherwise return non-empty collection.

But I am following that rule. `result` is set to null at the beginning so in all the early `return false` cases it will return null. Once I assign it an actual collection, I don't return without first checking if it's empty and if so, setting it back to null.